src/lib.rs

```diff
@@ -1,576 +1,333 @@
+            let expiry = &proof.payload.valid_until;
+            ensure!(
+                expiry.is_valid_at(now),
+                "invocation failed: proof expired at {expiry}"
+            );
+
+            // Verify that the capability is actually reached through:
+            let Some((vk, attenuation)) = proof.payload.attenuation.as_ref() else {
+                bail!("invocation failed: proof is missing delegation",);
+            };
+            ensure!(
+                vk == &self.identity,
+                "invocation failed: proof is missing delegation for capability of {}",
+                hex::encode(self.identity)
+            );
+
+            // Verify that the capability doesn't break out of attenuations:
+            ensure!(
+                capability.allowed_by_attenuation(attenuation),
+                "invocation failed"
+            );
+
+            // Continue checking the proof chain's integrity with this
+            // delegation's audience as the next issuer target:
+            current_issuer_target = audience;
+        }
 
-impl PartialOrd for VerifyingKeyWrapper {
-    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
-        Some(self.cmp(other))
-    }
-}
+        ensure!(
+        &invoker == current_issuer_target,
+        "invocation failed: expected delegation chain to end in the connection's owner {}, but the connection is authenticated by {} instead",
+        hex::encode(invoker),
+        hex::encode(current_issuer_target),
+    );
 
-impl Ord for VerifyingKeyWrapper {
-    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
-        self.0.as_bytes().cmp(other.0.as_bytes())
+        Ok(())
     }
 }
 
-impl<C: Capability> Caps<C> {
-    pub fn insert(&mut self, key: VerifyingKey, cap: C) {
-        self.0.insert(VerifyingKeyWrapper(key), cap);
-    }
-
-    pub fn get(&self, key: &VerifyingKey) -> Option<&C> {
-        self.0.get(&VerifyingKeyWrapper(*key))
-    }
-
-    pub fn iter(&self) -> impl Iterator<Item = (&VerifyingKey, &C)> {
-        self.0.iter().map(|(k, c)| (&k.0, c))
-    }
+/// A token for attenuated capability delegations
+#[derive(Clone, Serialize, Deserialize, Debug, PartialEq, Eq)]
+pub struct Rcan<A> {
+    /// The actual content.
+    pub payload: Payload<A>,
+    /// Signature over the serialized payload.
+    pub signature: Signature,
 }
 
 #[derive(Clone, Serialize, Deserialize, derive_more::Debug, PartialEq, Eq)]
-pub struct Payload<C: Capability> {
+pub struct Payload<A> {
     /// The issuer
     #[debug("{}", hex::encode(issuer))]
     issuer: VerifyingKey,
     /// The intended audience
     #[debug("{}", hex::encode(audience))]
     audience: VerifyingKey,
-    /// Nonce, to avoid replayability.
-    #[debug("{}", hex::encode(nonce))]
-    nonce: [u8; NONCE_SIZE],
-    /// Capabilities
-    #[serde(bound = "C: Capability")]
-    caps: Caps<C>,
+    /// Attenuation on delegated capabilities
+    attenuation: Option<(VerifyingKey, A)>,
```

Comment:
What does it mean when `attenuation` is empty?
The only reason I can come up with (we used this in the past with UCANs) is authenticating by signing a token with a challenge that's put into the nonce.
But there's no nonce in these tokens.

Reply:
yeah, good question, I wasn't sure if we should do this, before it was possible to have an empty map, but I guess it doesn't make much sense

Reply:
removed it